File: src/main/java/supercritical/api/unification/material/properties/CoolantProperty.java
package supercritical.api.unification.material.properties;

import net.minecraftforge.fluids.Fluid;

import gregtech.api.fluids.store.FluidStorageKey;
import gregtech.api.unification.material.Material;
import gregtech.api.unification.material.properties.IMaterialProperty;
import gregtech.api.unification.material.properties.MaterialProperties;
import gregtech.api.unification.material.properties.PropertyKey;
import lombok.Getter;
import lombok.Setter;
import supercritical.api.nuclear.fission.ICoolantStats;

public class CoolantProperty implements IMaterialProperty, ICoolantStats {

    @Setter
    @Getter
    private Material hotHPCoolant;
    @Setter
    @Getter
    private double moderatorFactor;
    /**
     * Roughly the heat transfer coefficient Do not put too much thought into this
     */
    @Setter
    @Getter
    private double coolingFactor;
    // in kelvin at standard conditions
    @Setter
    @Getter
    private double boilingPoint;
    // neutron absorption rate
    // in J/L
    @Setter
    @Getter
    private double heatOfVaporization;
    // in J/(kg*K)
    @Setter
    @Getter
    private double specificHeatCapacity;
    private boolean accumulatesHydrogen = false;

    @Setter
    @Getter
    private double slowAbsorptionFactor = 0;

    @Setter
    @Getter
    private double fastAbsorptionFactor = 0;
    // To store the specific key
    private final FluidStorageKey key;

    @Getter
    private final double mass;

    public CoolantProperty(Material mat, Material hotHPCoolant, FluidStorageKey key, double moderatorFactor,
                           double coolingFactor,
                           double boilingPoint, double heatOfVaporization,
                           double specificHeatCapacity) {
        this.hotHPCoolant = hotHPCoolant;
        this.moderatorFactor = moderatorFactor;
        this.coolingFactor = coolingFactor;
        this.boilingPoint = boilingPoint;
        this.heatOfVaporization = heatOfVaporization;
        this.specificHeatCapacity = specificHeatCapacity;
        this.key = key;
        this.mass = mat.getMass();
    }

    @Override
    public void verifyProperty(MaterialProperties properties) {
        properties.ensureSet(PropertyKey.FLUID, true);
    }

    public boolean accumulatesHydrogen() {
        return accumulatesHydrogen;
    }

    public CoolantProperty setAccumulatesHydrogen(boolean accumulatesHydrogen) {
        this.accumulatesHydrogen = accumulatesHydrogen;
        return this;
    }

    public Fluid getHotCoolant() {
        return hotHPCoolant.getFluid();
    }

    public CoolantProperty setFastAbsorptionFactor(double fastAbsorptionFactor) {
        this.fastAbsorptionFactor = fastAbsorptionFactor;
        return this;
    }

    public CoolantProperty setSlowAbsorptionFactor(double slowAbsorptionFactor) {
        this.slowAbsorptionFactor = slowAbsorptionFactor;
        return this;
    }

    public FluidStorageKey getCoolantKey() {
        return key;
    }
}

File: src/main/kotlin/supercritical/api/unification/material/properties/CoolantProperty.kt
package supercritical.api.unification.material.properties

import com.gregtechceu.gtceu.api.data.chemical.material.Material
import com.gregtechceu.gtceu.api.data.chemical.material.properties.FluidProperty
import com.gregtechceu.gtceu.api.data.chemical.material.properties.IMaterialProperty
import com.gregtechceu.gtceu.api.data.chemical.material.properties.MaterialProperties
import com.gregtechceu.gtceu.api.data.chemical.material.properties.PropertyKey
import com.gregtechceu.gtceu.api.fluids.store.FluidStorageKey
import net.minecraft.world.level.material.Fluid
import supercritical.api.nuclear.fission.ICoolantStats

class CoolantProperty(
    material: Material,
    var hotHPCoolant: Material,
    val coolantKey: FluidStorageKey?,
    override var moderatorFactor: Double,
    override var coolingFactor: Double,
    override var boilingPoint: Double,
    override var heatOfVaporization: Double,
    override var specificHeatCapacity: Double
) : IMaterialProperty, ICoolantStats {
    private var accumulatesHydrogen = false
    override var slowAbsorptionFactor = 0.0
    override var fastAbsorptionFactor = 0.0
    override val mass: Double
    override val hotCoolant: Fluid?
        get() = hotHPCoolant.fluid

    init {
        mass = material.mass.toDouble()
    }

    override fun verifyProperty(properties: MaterialProperties) {
        properties.ensureSet<FluidProperty?>(PropertyKey.FLUID, true)
    }

    fun setHotHPCoolant(hotHPCoolant: Material): CoolantProperty {
        this.hotHPCoolant = hotHPCoolant
        return this
    }

    fun setModeratorFactor(moderatorFactor: Double): CoolantProperty {
        this.moderatorFactor = moderatorFactor
        return this
    }

    fun setCoolingFactor(coolingFactor: Double): CoolantProperty {
        this.coolingFactor = coolingFactor
        return this
    }

    fun setBoilingPoint(boilingPoint: Double): CoolantProperty {
        this.boilingPoint = boilingPoint
        return this
    }

    fun setHeatOfVaporization(heatOfVaporization: Double): CoolantProperty {
        this.heatOfVaporization = heatOfVaporization
        return this
    }

    fun setSpecificHeatCapacity(specificHeatCapacity: Double): CoolantProperty {
        this.specificHeatCapacity = specificHeatCapacity
        return this
    }

    override fun accumulatesHydrogen(): Boolean {
        return accumulatesHydrogen
    }

    fun setAccumulatesHydrogen(accumulatesHydrogen: Boolean): CoolantProperty {
        this.accumulatesHydrogen = accumulatesHydrogen
        return this
    }

    fun setSlowAbsorptionFactor(slowAbsorptionFactor: Double): CoolantProperty {
        this.slowAbsorptionFactor = slowAbsorptionFactor
        return this
    }

    fun setFastAbsorptionFactor(fastAbsorptionFactor: Double): CoolantProperty {
        this.fastAbsorptionFactor = fastAbsorptionFactor
        return this
    }

}
